### PR TITLE
Update the image tag filtering for WP 6+

### LINKED ISSUE #219: WP 6.4 some content images are not cropped on the frontend
Certain images are not longer correctly cropped in the frontend of sites after WordPress 6.4 media changes. Simply downgrading to WP 6.3 is enough for the crop to be restored without other changes. 

The crop is showing correctly in the admin. It seems to affect portrait style images such as those on "meet the team" pages.

Debugging the code the image is not added to the selected images because it already has a srcset. If I bypass this check the presence of width and height attributes are also a problem for the correct processing of the image.

I've enclosed a visual diff of the same innerhtml for the image WP 6.3.1 and WP 6.4.3
image

I tried the `master` branch but there is no change.

This is the code going into https://github.com/humanmade/smart-media/blob/93961715c820dc8324946152d45039cac835d710/inc/cropper/namespace.php#L928

```
<img decoding="async" width="1920" height="2560" src="[https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=720%2C405&amp;crop_strategy=smart](https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=720%2C405&crop_strategy=smart)" alt="" class="wp-image-7335" srcset="[https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=1920%2C2560&amp;zoom=1](https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=1920%2C2560&zoom=1) 1920w, [https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=1920%2C2560&amp;zoom=0.5](https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=1920%2C2560&zoom=0.5) 960w, [https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=1920%2C2560&amp;zoom=0.25](https://kaplan.altis.dev/tachyon/sites/3/2024/03/IMG_3148-scaled.jpg?resize=1920%2C2560&zoom=0.25) 480w" sizes="(max-width: 1920px) 100vw, 1920px" />
```

### PR DESCRIPTION
**Pull request by roborourke**
Uses a less aggressive method of integrating with core's filtering of image content tags, prevents core's handling of width, height and srcset attributes in favour of using the content tag filter provided.

Fixes #219

### CODE REVIEW

File: inc/cropper/namespace.php [hunk 1]
```diff
@@ -884,143 +896,46 @@ function massage_meta_data_for_orientation( array $meta_data ) {
...
-		$context = current_filter();
+function content_img_tag( string $filtered_image, string $context, int $attachment_id ) : string {
+	if ( strpos( $filtered_image, TACHYON_URL ) === false ) {
+		return $filtered_image;
 	}
 
-	if ( false === strpos( $content, '<img' ) ) {
-		return $content;
-	}
+	$image_meta = wp_get_attachment_metadata( $attachment_id );
```
Comment (svandragt) [inline inc/cropper/namespace.php:911]: After this we can check if attachment_id === 0 then return filtered_image

File: inc/cropper/namespace.php [hunk 2]
```diff
@@ -884,143 +896,46 @@ function massage_meta_data_for_orientation( array $meta_data ) {
...
 
-	$images = Tachyon::parse_images_from_html( $content );
-	if ( empty( $images ) ) {
-		// No images, leave early.
-		return $content;
+	// Add 'width' and 'height' attributes if applicable.
+	if ( $attachment_id > 0 && ! str_contains( $filtered_image, ' width=' ) && ! str_contains( $filtered_image, ' height=' ) ) {
```
Comment (svandragt) [inline inc/cropper/namespace.php:914]: after following the advice above, we can remove the first check, simplifying the code.

File: inc/cropper/namespace.php [hunk 3]
```diff
@@ -884,143 +896,46 @@ function massage_meta_data_for_orientation( array $meta_data ) {
...
-			// Overwrite the ID when the same image is included more than once.
-			$attachment_ids[ $attachment_id ] = true;
-		}
-	}
-
-	if ( empty( $attachment_ids ) ) {
-		// No WP attachments, nothing further to do.
-		return $content;
+	// Add 'srcset' and 'sizes' attributes if applicable.
+	if ( $attachment_id > 0 && ! str_contains( $filtered_image, ' srcset=' ) ) {
```
Comment (svandragt) [inline inc/cropper/namespace.php:919]: same here

File: inc/cropper/namespace.php [hunk 4]
```diff
@@ -884,143 +896,46 @@ function massage_meta_data_for_orientation( array $meta_data ) {
...
+ * @param string $filtered_image Full img tag with attributes that will replace the source img tag.
+ * @param string $context Additional context, like the current filter name or the function name from where this was called.
+ * @param int $attachment_id  The image attachment ID. May be 0 in case the image is not an attachment.
+ * @return string Full img tag with attributes that will replace the source img tag.
  */
-function filter_content_tags( string $content, string $context = null ) : string {
-	if ( null === $context ) {
-		$context = current_filter();
+function content_img_tag( string $filtered_image, string $context, int $attachment_id ) : string {
+	if ( strpos( $filtered_image, TACHYON_URL ) === false ) {
```
Comment (svandragt) [inline inc/cropper/namespace.php:907]: Doesn't this create a hard requirement on tachyon? What if people use this plugin without it? 

The readme says:

> Some features in this plugin will work on their own however some are designed to augment the existing tools we use such as [Tachyon](https://github.com/humanmade/tachyon).

So I think let's check here if tachyon is on the system

File: inc/cropper/namespace.php [hunk 5]
```diff
@@ -884,143 +896,46 @@ function massage_meta_data_for_orientation( array $meta_data ) {
...
+ * We override this to ensure compatibility with Tachyon & smart media.
+ *
+ * @param bool $value The filtered value, defaults to <code>true</code>.
+ * @param string $image The HTML <code>img</code> tag where the attribute should be added.
+ * @param string $context Additional context about how the function was called or where the img tag is.
+ * @param int $attachment_id The image attachment ID.
+ * @return bool The filtered value, defaults to <code>true</code>.
+ */
+function img_tag_add_attr( bool $value, string $image ) : bool {
+	return strpos( $image, TACHYON_URL ) === false ? $value : false;
```
Comment (svandragt) [inline inc/cropper/namespace.php:938]: This would also need a check to see if TACHYON_URL is defined.

File: plugin.php [hunk 6]
```diff
@@ -4,7 +4,7 @@
  * Description: Advanced media tools that take advantage of Rekognition and Tachyon.
  * Author: Human Made Limited
  * License: GPL-3.0
- * Version: 0.4.3
+ * Version: 0.4.6
```
Comment (svandragt) [inline plugin.php:7]: This is not backwards compatible as we're removing and adding filters so I would have called it 0.5.0, but I don't know the project so feel free to ignore.

**Review by svandragt — CHANGES_REQUESTED** [review]
Great work, just suggesting a few improvements to make it more widely useful and maintainable.

File: plugin.php (hunk 6 above)
Comment (roborourke) [reply plugin.php:7]: I was thinking about that but it is backwards compatible in Altis terms. I just checked when 0.4.0 of this plugin was released though and that was in 2021 before WP 6.0 so you're right, it'll just take a bit longer to get the release out as we'll need an Altis patch too.

File: inc/cropper/namespace.php (hunk 2 above)
Comment (roborourke) [reply inc/cropper/namespace.php:914]: core copy/pasta here, you're right it could simplified a bit.

File: plugin.php (hunk 6 above)
Comment (svandragt) [reply plugin.php:7]: ah, well this is no longer an altis plugin.

**Pull request merged**